File: libraries/apollo-runtime-java/src/main/java/com/apollographql/apollo3/runtime/java/NetworkTransport.java
package com.apollographql.apollo3.runtime.java;

import com.apollographql.apollo3.api.ApolloRequest;
import com.apollographql.apollo3.api.Operation;
import com.apollographql.apollo3.runtime.java.interceptor.ApolloDisposable;
import org.jetbrains.annotations.NotNull;

public interface NetworkTransport {
  <D extends Operation.Data> void execute(@NotNull ApolloRequest<D> request, @NotNull ApolloCallback<D> callback, ApolloDisposable disposable);
}

File: libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/NetworkTransport.kt
package com.apollographql.apollo.network

import com.apollographql.apollo.api.ApolloRequest
import com.apollographql.apollo.api.ApolloResponse
import com.apollographql.apollo.api.Operation
import kotlinx.coroutines.flow.Flow

interface NetworkTransport {

  /**
   * Execute a request.
   *
   * This function must not throw in normal circumstances. Instead, emit an [ApolloResponse] with a non null [ApolloResponse.exception].
   * It may throw if the error is the result of a programming error like wrongly configured interceptors.
   *
   * @param request the request to execute.
   * @return a flow of responses.
   */
  fun <D : Operation.Data> execute(
      request: ApolloRequest<D>,
  ): Flow<ApolloResponse<D>>

  fun dispose()
}
